SimCalorimetry/HGCalAssociatorProducers/plugins/LayerClusterAssociatorByEnergyScoreImpl.cc

```diff
@@ -0,0 +1,531 @@
+  }
+  nCaloParticles = cPIndices.size();
+
+  // Initialize cPOnLayer. To be returned outside, since it contains the
+  // information to compute the CaloParticle-To-LayerCluster score.
+  hgcal::caloParticleToLayerCluster cPOnLayer;
+  cPOnLayer.resize(nCaloParticles);
+  for (unsigned int i = 0; i < nCaloParticles; ++i) {
+    cPOnLayer[i].resize(layers_ * 2);
+    for (unsigned int j = 0; j < layers_ * 2; ++j) {
+      cPOnLayer[i][j].caloParticleId = i;
+      cPOnLayer[i][j].energy = 0.f;
+      cPOnLayer[i][j].hits_and_fractions.clear();
+      //cPOnLayer[i][j].layerClusterIdToEnergyAndScore.reserve(nLayerClusters); // Not necessary but may improve performance
+    }
+  }
+
+  // Fill detIdToCaloParticleId_Map and update cPOnLayer
+  std::unordered_map<DetId, std::vector<hgcal::detIdInfoInCluster>> detIdToCaloParticleId_Map;
+  for (const auto& cpId : cPIndices) {
+    const SimClusterRefVector& simClusterRefVector = caloParticles[cpId].simClusters();
+    for (const auto& it_sc : simClusterRefVector) {
+      const SimCluster& simCluster = (*(it_sc));
+      const auto& hits_and_fractions = simCluster.hits_and_fractions();
+      for (const auto& it_haf : hits_and_fractions) {
+        DetId hitid = (it_haf.first);
+        int cpLayerId = recHitTools_->getLayerWithOffset(hitid) + layers_ * ((recHitTools_->zside(hitid) + 1) >> 1) - 1;
+        std::map<DetId, const HGCRecHit*>::const_iterator itcheck = hitMap_->find(hitid);
+        if (itcheck != hitMap_->end()) {
+          const HGCRecHit* hit = itcheck->second;
+          auto hit_find_it = detIdToCaloParticleId_Map.find(hitid);
+          if (hit_find_it == detIdToCaloParticleId_Map.end()) {
+            detIdToCaloParticleId_Map[hitid] = std::vector<hgcal::detIdInfoInCluster>();
+            detIdToCaloParticleId_Map[hitid].emplace_back(hgcal::detIdInfoInCluster{cpId, it_haf.second});
+          } else {
+            auto findHitIt = std::find(detIdToCaloParticleId_Map[hitid].begin(),
+                                       detIdToCaloParticleId_Map[hitid].end(),
+                                       hgcal::detIdInfoInCluster{cpId, it_haf.second});
+            if (findHitIt != detIdToCaloParticleId_Map[hitid].end()) {
+              findHitIt->fraction += it_haf.second;
+            } else {
+              detIdToCaloParticleId_Map[hitid].emplace_back(hgcal::detIdInfoInCluster{cpId, it_haf.second});
+            }
+          }
+          cPOnLayer[cpId][cpLayerId].energy += it_haf.second * hit->energy();
+          // We need to compress the hits and fractions in order to have a
+          // reasonable score between CP and LC. Imagine, for example, that a
+          // CP has detID X used by 2 SimClusters with different fractions. If
+          // a single LC uses X with fraction 1 and is compared to the 2
+          // contributions separately, it will be assigned a score != 0, which
+          // is wrong.
+          auto& haf = cPOnLayer[cpId][cpLayerId].hits_and_fractions;
```

Comment:
const

Reply:
@lecriste this comment should still be addressed

Reply:
@kpedro88 this `haf` can not be `const` because it may be modified 4 lines later:
https://github.com/cms-sw/cmssw/pull/29844/files/f12af7c4c5e4912961e408b0e4bdc3cf4e720e54#diff-93cdc624d14004b424f0e32ccd9df70aR95

Reply:
Ah, I see, `found` is a member of that collection. Thanks for pointing this out.